src/G2/Equiv/Verifier.hs

```diff
@@ -145,16 +122,69 @@ stackWrap sk e =
 loc_name :: Name
 loc_name = Name (DT.pack "STACK") Nothing 0 Nothing
 
+rec_name :: Name
+rec_name = Name (DT.pack "REC") Nothing 0 Nothing
+
+wrapRecursiveCall :: Name -> Expr -> Expr
+-- TODO attempt to prevent double wrapping
+wrapRecursiveCall n e@(Tick (NamedLoc n'@(Name t _ _ _)) e') =
+  if t == DT.pack "REC"
+  then e
+  else Tick (NamedLoc n') $ wrcHelper n e'
+wrapRecursiveCall n e@(Var (Id n' _)) =
+  if n == n'
+  then Tick (NamedLoc rec_name) e
+  else wrcHelper n e
+wrapRecursiveCall n e = wrcHelper n e
+
+wrcHelper :: Name -> Expr -> Expr
+wrcHelper n = modifyChildren (wrapRecursiveCall n)
+
+-- do not allow wrapping for symbolic variables
+recWrap :: ExprEnv -> Name -> Expr -> Expr
+recWrap h n =
+  if E.isSymbolic n h
+  then id
+  else (wrcHelper n) . (wrapRecursiveCall n)
+
+-- first Name is the one that maps to the Expr in the environment
+-- second Name is the one that might be wrapped
+wrapIfCorecursive :: G.CallGraph -> ExprEnv -> Name -> Name -> Expr -> Expr
+wrapIfCorecursive cg h n m e =
+  let n_list = G.reachable n cg
+      m_list = G.reachable m cg
+  in
+  if (n `elem` m_list) && (m `elem` n_list)
+  then recWrap h m e
+  else e
+
+-- the call graph must be based on the given environment
+-- the Name must map to the Expr in the environment
+wrapAllRecursion :: G.CallGraph -> ExprEnv -> Name -> Expr -> Expr
+wrapAllRecursion cg h n e =
+  let n_list = G.reachable n cg
+  in
+  if (not $ E.isSymbolic n h) && (n `elem` n_list)
+  then foldr (wrapIfCorecursive cg h n) e n_list
+  else e
+
 tickWrap :: Expr -> Expr
 tickWrap e = Tick (NamedLoc loc_name) e
 
 exprWrap :: Stck.Stack Frame -> Expr -> Expr
 exprWrap sk e = stackWrap sk $ tickWrap e
 
+-- TODO a Var can satisfy EVF if it's symbolic or if it's unmapped
```

Comment:
Maybe just stick a comment in here to explain why we don't use isExprValueForm directly